UsbIpServer/RegistryUtils.cs

```diff
@@ -5,45 +5,207 @@
 using System;
 using System.Linq;
 using System.Security.Principal;
+using System.Globalization;
 using Microsoft.Win32;
+using System.Collections.Generic;
 
 namespace UsbIpServer
 {
     static class RegistryUtils
     {
         const string DevicesRegistryPath = @"SOFTWARE\usbipd-win";
 
-        public static bool IsDeviceAvailable(string busId)
+        static class DeviceFilter
         {
-            return Registry.LocalMachine.OpenSubKey(DevicesRegistryPath)?.GetSubKeyNames().Any(x => x == busId) ?? false;
+            public const string VENDOR_ID = "VendorId";
+            public const string PRODUCT_ID = "ProductId";
+            public const string BCD_DEVICE = "BcdDevice";
+            public const string DEVICE_CLASS = "DeviceClass";
+            public const string DEVICE_SUBCLASS = "DeviceSubClass";
+            public const string DEVICE_PROTOCOL = "DeviceProtocol";
+            public const string DEV_NUM = "DevNum";
+            public const string BUS_ID = "BusId";
         }
 
-        public static void SetDeviceAvailability(string busId, bool enable)
+        public static bool IsDeviceShared(ExportedDevice device)
         {
-            if (enable)
-            { 
-                Registry.LocalMachine.CreateSubKey($@"{DevicesRegistryPath}\{busId}");
-            } else if (IsDeviceAvailable(busId))
+            return GetRegistryKey(device) != null;
+        }
+
+        public static bool IsDeviceAttached(ExportedDevice device)
+        {
+            var key = GetRegistryKey(device);
+            if (key != null)
+            {
+                return key.GetSubKeyNames().Contains("Attached");
+            }
+
+            return false;
+        }
+
+        static bool IsDeviceMatch(RegistryKey deviceKey, ExportedDevice device)
+        {
+            
+            var cultureInfo = new CultureInfo("en-US");
```

Comment:
That's a bit heavy. `CultureInfo.InvariantCulture` is perfectly fine here. It's not that we are formatting numbers, currency, dates, or times here.

Reply:
Done